Fix CNP/CCNP update when selectors change from nil to empty non-nil slices

pippolo84 opened the pull request:
As suggested in https://github.com/cilium/cilium/issues/33432, the autogenerated DeepEqual methods to compare `IngressCommonRule` structs and `EgressCommonRule` structs do not take into account that a nil slice in one of the fields should not be compared equal to a empty non-nil slice in the same field.
Since https://github.com/cilium/cilium/pull/29608, the semantic of a empty non-nil slice in both Ingress and Egress section of CNPs and CCNPs has been changed to not select any identity.
Therefore, a CNP with a nil ToEndpoints should not compare equal to a CNP with an empty non-nil ToEndpoints fields, otherwise the following update:

```
apiVersion: "cilium.io/v2"
kind: CiliumNetworkPolicy
metadata:
  name: test
spec:
  endpointSelector:
    matchLabels:
      run: tmp-shell
  egress:
  - toCIDR:
    - 1.1.1.1/32
    toPorts:
      - ports:
          - port: "80"
```

-->

```
apiVersion: "cilium.io/v2"
kind: CiliumNetworkPolicy
metadata:
  name: test
spec:
  endpointSelector:
    matchLabels:
      run: tmp-shell
  egress:
  - toCIDR:
    - 1.1.1.1/32
    toEndpoints: [] # <-- this
    toPorts:
      - ports:
          - port: "80"
```

is incorrectly discarded, instead of being propagated to the policy repository.

Related: https://github.com/cilium/cilium/pull/29608
Fixes: https://github.com/cilium/cilium/issues/33432

pippolo84 committed:
policy/api: Use a custom DeepEqual method to compare IngressCommonRule

The semantic of a nil slice in one of the IngressCommonRule fields is
different from the semantic of an empty non-nil slice. In order to
compare two IngressCommonRule instances correctly, we add a custom
DeepEqual method the explicitly checks for this case before calling the
autogenerated method.

This allows to correctly propagate CNP/CCNP updates when, for example,
the FromEndpoints selector is changed from nil to an empty non-nil
slice. In the latter case, the CNP should not select any identity,
falling back to the default behavior (e.g: default deny for an allow
policy).

Fixes: e97df7badd ("policy: Do not select any identity with ingress empty slices")
Fixes: https://github.com/cilium/cilium/issues/33432

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

pippolo84 committed:
policy/api: Use a custom DeepEqual method to compare EgressCommonRule

The semantic of a nil slice in one of the EgressCommonRule fields is
different from the semantic of an empty non-nil slice. In order to
compare two EgressCommonRule instances correctly, we add a custom
DeepEqual method the explicitly checks for this case before calling the
autogenerated method.

This allows to correctly propagate CNP/CCNP updates when, for example,
the ToEndpoints selector is changed from nil to an empty non-nil
slice. In the latter case, the CNP should not select any identity,
falling back to the default behavior (e.g: default deny for an allow
policy).

Fixes: 6ccd044b05 ("policy: Do not select any identity with egress empty slices")
Fixes: https://github.com/cilium/cilium/issues/33432

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

christarazi reviewed (commented):
Nice find, LGTM. Could you add a unit test similar to https://github.com/cilium/cilium/pull/33439 to actually test the expected behavior end to end?

pippolo84 committed:
policy/k8s: Add GH issue 33432 non regression test

Add a non regression test for GitHub issue #33432: updating a CNP with a
nil ToEndpoints slice to an empty non-nil ToEndpoints slice should
result in the update not being discarded and the rules in the policy
repository reporting the empty non-nil ToEndpoints slice.

Related: https://github.com/cilium/cilium/issues/33432

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

derailed commented on pkg/policy/api/ingress.go:
```diff
@@ -99,6 +101,31 @@ type IngressCommonRule struct {
+// of an empty non-nil slice, thus it explicitly checks for that case before calling
+// the autogenerated method.
+func (in *IngressCommonRule) DeepEqual(other *IngressCommonRule) bool {
+	if in.FromEndpoints == nil && other.FromEndpoints != nil ||
```
This seems very repetitive. Perhaps a convenience  aka "isNilSlice" could help?

pippolo84 replied:
I've added a `XorNil` helper in `pkg/slices`, see https://github.com/cilium/cilium/pull/33506/commits/f2301b56451d96143644ec680be55fd0a9962de8

derailed replied:
Brilliant! Thank you Fabio!

derailed reviewed (approved):
@pippolo84 Great catch Fabio!

pippolo84 committed:
slices: Add XorNil helper

XorNil is a helper that returns true only if one of the two input slices
is nil and the other is not.

It is useful when comparing two slices and the semantic of a nil slice
is different from the semantic of an empty non-nil one.

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>
